pkg/activator/net/throttler.go

```diff
@@ -0,0 +1,489 @@
+	logger *zap.SugaredLogger) *revisionThrottler {
+	var revBreaker breaker
+	if containerConcurrency == 0 {
+		revBreaker = NewInfiniteBreaker()
+	} else {
+		revBreaker = queue.NewBreaker(breakerParams)
+	}
+	return &revisionThrottler{
+		revID:                revID,
+		containerConcurrency: containerConcurrency,
+		breaker:              revBreaker,
+		logger:               logger,
+	}
+}
+
+// Returns clusterIP if it is the current valid dest. If neither clusterIP or podIPs are valid
+// dests returns error.
+func (rt *revisionThrottler) checkClusterIPDest() (string, error) {
+	rt.mux.RLock()
+	defer rt.mux.RUnlock()
+
+	if rt.clusterIPDest == "" && len(rt.podIPTrackers) == 0 {
+		return "", errors.New("made it through breaker but we have no clusterIP or podIPs. This should" +
+			" never happen" + rt.revID.String())
+	}
+
+	return rt.clusterIPDest, nil
+}
+
+// Returns a dest after incrementing it's request cnt and a completion callback
+// to be called after request completion
+func (rt *revisionThrottler) acquireDest() (string, func()) {
+	rt.mux.Lock()
+	defer rt.mux.Unlock()
+
+	if rt.clusterIPDest != "" {
+		return rt.clusterIPDest, func() {}
+	}
+
+	var leastConn *podIPTracker
+	// Find the dest with fewest active connections.
+	for _, tracker := range rt.podIPTrackers {
+		if leastConn == nil || leastConn.requests > tracker.requests {
+			leastConn = tracker
+		}
+	}
+
+	if leastConn != nil {
+		leastConn.requests++
+		return leastConn.dest, func() {
+			rt.mux.Lock()
+			defer rt.mux.Unlock()
+			leastConn.requests--
+		}
+	}
+	return "", func() {}
+}
+
+func (rt *revisionThrottler) try(ctx context.Context, function func(string) error) error {
+	var ret error
+
+	rt.logger.Debug("Trying")
+	if !rt.breaker.Maybe(ctx, func() {
+		// See if we can get by with only a readlock
+		dest, err := rt.checkClusterIPDest()
+		if err != nil {
+			ret = err
+			return
+		}
+
+		if dest != "" {
+			ret = function(dest)
+			return
+		}
+
+		// Try again with a write lock falling back to a podIP dest
+		dest, completionCb := rt.acquireDest()
+		defer completionCb()
+
+		if dest == "" {
+			ret = errors.New("no podIP destination found, this should never happen")
+		} else {
+			ret = function(dest)
+		}
+	}) {
+		return fmt.Errorf("revision %q exceeded capacity, failing request", rt.revID.String())
+	}
+	return ret
+}
+
+func (rt *revisionThrottler) calculateCapacity(size, activatorCount, maxConcurrency int) int {
+	targetCapacity := int(rt.containerConcurrency) * size
+
+	if size > 0 && (rt.containerConcurrency == 0 || targetCapacity > maxConcurrency) {
+		// If cc==0, we need to pick a number, but it does not matter, since
+		// infinite breaker will dole out as many tokens as it can.
+		targetCapacity = maxConcurrency
+	} else if targetCapacity > 0 {
+		targetCapacity = minOneOrValue(targetCapacity / minOneOrValue(activatorCount))
+	}
+
+	return targetCapacity
+}
+
+func (rt *revisionThrottler) updateCapacity(throttler *Throttler, backendCount int) {
```

Comment:
This feels like a layering problem. Why not just pass in the `activatorCount` and `maxconcurrency`?